cmd/nvidia-dra-controller/imex.go

```diff
@@ -102,35 +118,50 @@ func StartIMEXManager(ctx context.Context, config *Config) (*ImexManager, error)
 }
 
 // manageResourceSlices reacts to added and removed IMEX domains and triggers the creation / removal of resource slices accordingly.
-func (m *ImexManager) manageResourceSlices(ctx context.Context, owner resourceslice.Owner, addedDomainsCh <-chan string, removedDomainsCh <-chan string) error {
-	driverResources := &resourceslice.DriverResources{}
-	controller, err := resourceslice.StartController(ctx, m.clientset, DriverName, owner, driverResources)
+func (m *ImexManager) manageResourceSlices(ctx context.Context) error {
+	klog.Info("Start streaming IMEX domains from nodes...")
+	addedDomainsCh, removedDomainsCh, err := m.streamImexDomains(ctx)
+	if err != nil {
+		return fmt.Errorf("error streaming IMEX domains: %w", err)
+	}
+
+	klog.Info("Start publishing IMEX channels to ResourceSlices...")
+	controller, err := resourceslice.StartController(ctx, m.clientset, m.driverName, m.owner, m.driverResources)
 	if err != nil {
 		return fmt.Errorf("error starting resource slice controller: %w", err)
 	}
 
-	imexDomainOffsets := new(imexDomainOffsets)
 	m.waitGroup.Add(1)
 	go func() {
 		defer m.waitGroup.Done()
 		for {
 			select {
 			case addedDomain := <-addedDomainsCh:
-				offset, err := imexDomainOffsets.add(addedDomain, ResourceSliceImexChannelLimit, DriverImexChannelLimit)
-				if err != nil {
-					klog.Errorf("Error calculating channel offset for IMEX domain %s: %v", addedDomain, err)
-					return
-				}
 				klog.Infof("Adding channels for new IMEX domain: %v", addedDomain)
-				driverResources := driverResources.DeepCopy()
-				driverResources.Pools[addedDomain] = generateImexChannelPool(addedDomain, offset, ResourceSliceImexChannelLimit)
-				controller.Update(driverResources)
+				if err := m.addImexDomain(addedDomain); err != nil {
+					klog.Errorf("Error adding channels for IMEX domain %s: %v", addedDomain, err)
+					if errors.As(err, &transientError{}) {
```

Comment:
As per https://pkg.go.dev/errors#As , I think we want `if !errors.As(err, &transientError{})`

Reply:
IIUC as is we will only retry if `errors.as` return true

Reply:
Yes, which is exactly what we want. We only want to retry if the error is a transient error, which it will only be for the one error case we have marked as a transient error, i.e. the case where we have exceeded the limit on the available offsets to assign.

Reply:
now makes sense, I was looking at it from the other way around. I was not understanding the approach